Use redux to list tasks

probablycorey opened the pull request:
This is a SUPER WIP for adding redux. I'm just opening the PR now to give visibility to me working on it. When I'm ready for review I'll write a new comment and update this body.

Closes #4

Commit:
Add basic reducer and test

Commit:
Add redux and it’s helper packages

Commit:
Start using redux with react

Commit:
Create TASK_ADD action

Commit:
Use redux in App

Commit:
Update tests

Commit:
Show tasks in task list

Commit:
I guess I need to use Provider!

Commit:
Ignore tasks

Commit:
Don’t show archived tasks

Commit:
Don’t show snoozed tasks

Commit:
Show archived tasks that have been updated

Commit:
Make test names more readable

cheshire137 commented on components/TaskList/TaskList.jsx:
```diff
@@ -1,20 +1,42 @@
-    )
+const TaskList = ({tasks}) => {
+  const taskElements = tasks
+    .filter((task) => {
```
Oh no! The period-on-the-next-line strikes again!

J/k, BTW, it's cool.

probablycorey replied:
I'm following the AirBnb style guide!

`Use indentation when making long method chains. Use a leading dot, which emphasizes that the line is a method call, not a new statement.`

cheshire137 commented on components/TaskList/TaskList.jsx:
```diff
@@ -1,20 +1,42 @@
+
+const mapStateToProps = (tasks) => {
+  return {
+    tasks: tasks
```
Our linter might fuss at you for this. You should be able to do `{ tasks }`.

probablycorey replied:
The linter is very unhappy with me right now.

cheshire137 commented on test/components/TaskList-test.js:
```diff
@@ -0,0 +1,47 @@
+    assert(element)
+  })
+
+  it('do not show tasks that are ignored, archived or snoozed', function() {
```
It do not. Maybe `does`? :D

Commit:
Update tests

Commit:
Merge remote-tracking branch 'origin/master' into list-tasks

# Conflicts:
#	components/TaskList/TaskList.jsx
#	package.json
#	test/components/app-test.js

Commit:
Make tests run

Commit:
Use better test names

Commit:
Not an atomic commit

Commit:
Get rid of this rogue reducer!

Commit:
Add Task

Commit:
Set the key correctly

Commit:
Move Task creation inline

Commit:
Get rid of mapDispatchToProps

Commit:
Don’t let reducers mutate state!

Commit:
I have no idea why this was here

Commit:
Fix linter errors

Commit:
Fix Task so snoozing works

Commit:
Use big letters

Commit:
Fix linting errors

Commit:
Don’t give default tasks

Commit:
Make tests work

cheshire137 commented on src/components/Task/Task.jsx:
```diff
@@ -0,0 +1,41 @@
+class Task extends React.Component {
+  onChange(event) {
+    const { task } = this.props
+    const type = event.target.checked ? 'TASKS_SELECT' : 'TASKS_UNSELECT'
```
Maybe `DESELECT`?

probablycorey replied:
I don't really like UNSELECT or DESELECT, but I'll go with DESELECT for now

cheshire137 commented on src/components/Task/Task.jsx:
```diff
@@ -0,0 +1,41 @@
+    )
+  }
+}
+module.exports = connect()(Task)
```
Ooh strange.

cheshire137 commented on src/components/Task/Task.jsx:
```diff
@@ -0,0 +1,41 @@
+const React = require('react')
```
Instead of adding this new class/file, you could overwrite TaskListItem that got added in https://github.com/cheshire137/gh-notifications-snoozer/pull/16.

cheshire137 commented on src/components/TaskList/TaskList.jsx:
```diff
@@ -1,20 +1,31 @@
-module.exports = TaskList
+const mapStateToProps = tasks => ({ tasks })
+
+module.exports = connect(mapStateToProps)(TaskList)
```
Can you explain what's happening with lines 29-31?

probablycorey replied:
The best way is to read http://redux.js.org/index.html :wink: but I'll try a tl;dr.

With Redux, an app has one object that holds every bit of state. You can pass this state from component to component via props, but that gets a little cumbersome. So in redux people use `connect`. It gives a component access to the global state by adding it to the components props. If you don't want to give the component access to the entire state, you can reduce what it has access to using `mapStateToProps`.

After explaining that I realize I don't need to use mapStateToProps here, so I'm going to remove it.

probablycorey replied:
> After explaining that I realize I don't need to use mapStateToProps here, so I'm going to remove it.

I'm actually going to leave this as is for now since I can't figure out how to get the state out of the props without using `mapStateToProps`

cheshire137 commented on src/reducers/reducer.js:
```diff
@@ -0,0 +1,39 @@
+    case 'TASKS_UNSELECT':
+      return tasks.map(task => {
+        if (task.id === action.task.id) {
+          return Object.assign({}, task, { isSelected: false })
```
What's this doing?

probablycorey replied:
Redux has three main rules http://redux.js.org/docs/introduction/ThreePrinciples.html

Because state can't be mutated, this creates a new object based on the original task and adds the `isSelected` field. It is similar to underscore's `extend` function.

cheshire137 commented on test/components/TaskList-test.js:
```diff
@@ -0,0 +1,61 @@
+    assert.equal(2, element.querySelectorAll("li").length)
+  })
+
+  it('unarchives a task if the takes has updates', function() {
```
``` diff
- takes
+ task
```

cheshire137 commented on test/components/TaskList-test.js:
```diff
@@ -0,0 +1,61 @@
+
+  it('unarchives a task if the takes has updates', function() {
+    const tasks = defaultTasks.map(task => {
+      task.updateAt = new Date()
```
Is `updateAt` the field name we'll get back from the GitHub API? I think it uses snake_case, at least results from the /notifications endpoint.

probablycorey replied:
For now it is using the dummy data found in https://github.com/cheshire137/gh-notifications-snoozer/pull/14/files#diff-053da0fe3233c4c7a2ad76216a682b65R2.

I want to get the real endpoints hooked up in a new PR. When we do I'd like to force the GitHub API result objects into a camelcase format so it looks like regualr javascript.

cheshire137 commented on test/components/TaskList-test.js:
```diff
@@ -0,0 +1,61 @@
+
+  context('when the snooze button is pressed', function() {
+    it('snoozes the selected tasks', function() {
+      const snoozeButton = element.querySelector('button#snooze')
```
Could you use `element.getElementById('snooze')`?

probablycorey replied:
I got rid of all of the component tests because I'm not sure of the best way to test them right now.

I think I'll try and use https://github.com/airbnb/enzyme to test components in a follow up PR. It was the recommended framework to use in the React Docs https://facebook.github.io/react/docs/test-utils.html.

cheshire137 commented on src/reducers/Tasks-test.js:
```diff
@@ -1,21 +0,0 @@
-const assert = require('assert')
```
Re: the commit message

![dog-no-idea](https://cloud.githubusercontent.com/assets/82317/17075049/d116895e-504d-11e6-9219-88425421346c.jpg)

Commit:
Get rid of the annoying npm error

Commit:
Get tests working

Commit:
I’m not happy with how the tests worked

So I’ve just stubbed them for now.

Commit:
Merge remote-tracking branch 'origin/master' into list-tasks

# Conflicts:
#	src/components/TaskList/TaskList.jsx
#	test/components/App-test.js

Commit:
Don’t use notifications yet

probablycorey commented:
I think getting this PR in and getting the notifications changes in there might become too big. So I'm going to add the actual notification data in a new PR and update this to use dummy data. It will be easier to explain that way.

Commit:
Fix linter errors

Commit:
Add dummy data

Commit:
Make arrow-body-style a warning

Commit:
Don’t actually write the reducer tests yet

Commit:
Remove verify-token

@cheshire137 this wasn’t working anymore because it was a pure node script and the GitHub class was using some syntax only available to babel. So I just removed this for now.

Commit:
Use TASKS_DESELECT

probablycorey commented:
Ok, this is now ready. It is mostly just boilerplate for future usage, but I would rather keep the PR simple and put more logic into followup PRs. If this looks good to you, we can follow it up with:
1. A PR that uses the GitHub API to populate the tasks
2. A PR that uses enzyme to test components
3. A PR that uses [actions](http://redux.js.org/docs/basics/Actions.html) instead of raw calls to `dispatch`
4. A PR that adds tests the actions and reducer

cheshire137 commented on test/components/TaskList-test.js:
```diff
@@ -0,0 +1,20 @@
+
+  context('when the archive button is clicked', () => {
+    it('hides selected tasks')
+    it('updates the selected task\'s `archived_at` field')
```
Fear not the double quote!

cheshire137 commented:
Here we go!

probablycorey commented:
💥
